tests/ctst/HOW_TO_WRITE_TESTS.md

```diff
@@ -0,0 +1,127 @@
+harder debugging, and longer test execution.
+
+Examples:
+
+- Location creation;
+- Bucket Website endpoints.
+
+... must be created before the test start, or in a dedicated set of tests,
+that is, a set of test executed before all tests having a specific tag
+used for identifying the feature(s).
+
+**Rule #4**
+
+> Do not use `atMostOnePicklePerTag`.
+
+If a set of scenario requires the use of `atMostOnePicklePerTag`, they won't
+be executed in parallel, which is:
+
+- Not realistic for the production environment.
+- Not efficient for the test execution: the duration of tests will suffer.
+- A proof that the test is not following **Rule #1**.
+
+Ensuring this rule may be hard, and idempotent scenarios are not always
+possible. Solutions exist:
+
+- Pre-create the resources required for the tests before they start.
+- Use `Before` / `After` hooks to setup your environment once.
+  Concurrency can be controlled with lock files.
+- Make your assertions more flexible: strict and absolute checks are good
+  for unit or functional testing, but in integration tests, prefer using
+  relative checks.
+- As a last resort, we might have a dedicated test suite.
+
+**Rule #5**
+
+> Focus on validating features.
+
+It is tempting to write scenarios that test more than what is needed
+at the acceptance testing layer: please use unit and functional tests to fully
+validate the behavior of each component; and end-to-end tests only to ensure
+that the product-level feature is working as expected per the requirements.
+Making good use of unit and functional tests makes testing all cases, including
+corner cases or error handling, much easier, leading to higher quality; and
+improves velocity by having a much faster feedback loop.
+
+**Rule #6**
+
+> Ensure assertions are idempotent after sleep times.
+
+Some tests might need to "sleep" for some time, either static or using
+more complex logic. The subsequent assertions must take into account
+any operation on the platform that might be asynchronous and take place
+during the sleep time.
+
+**Rule #7**
+
+> Avoid using the @Flaky tag.
+
+Whenever there is a flaky, there is an issue : either a bug (corner case or
+race condition) in the product, or simply an issue in the test. So the
+issue should be fixed either way. More often than not, retrying the build
+(i.e. ignoring the flaky) or automating the retry with @Flaky is just
+masking a bug, and leaving it in production instead of fixing it : so it
+must not be done. One frequent problem in flaky tests is that they are not
+idempotent, or that we did not do our job correctly to understand the
+problem. Flaky tests must be investigated, and issues in the tests must be
+fixed to ensure we don't ignore an actual problem in the product.
+
+**Rule #8**
+
+> Teardown resources.
+
+If a test creates resources, ensure they are deleted at the end of the test.
+Do not delete the resources in case of error, to ease debugging.
```

Comment:
Thank you for all the feedback I updated the page accordingly - I agree with everything but this on. I updated as you suggested though.

I think we should not cleanup resources in case of error even if we're not running debug. The reason is that as each test uses a partitioned set of resources (bucket/objects, possibly account), not cleaning in case of error (or flaky) helps us finding the root cause much more easily, by having access to the metadata ind the mongodb dumps, as well as the logs of all operations done by CTST for the failing test (identifiable with a single id). The overhead is minimal and it should not affect other tests in any way, even if the bucket is still here.